File: day-24/lobby_layout.py
import re


input_file = 'day-24/input.txt'

#Use cube co-ordinates system:https://www.redblobgames.com/grids/hexagons/ CF 2017:11
deltas = dict(zip(['e','ne','nw','w','se','sw'], [(1, -1, 0), (1, 0, -1), (0, 1, -1), (-1, 1, 0), (0, -1, 1), (-1, 0, 1)]))

def convert_line(line):
  return re.findall('('+'|'.join(deltas.keys())+')',line)

def get_initial_flipped(input):
  flipped_tiles = set()
  for line in input.splitlines():
    x,y,z = 0,0,0
    directions = convert_line(line)
    for dir in directions:
      x += deltas[dir][0]
      y += deltas[dir][1]
      z += deltas[dir][2]
    if (x,y,z) in flipped_tiles:
      flipped_tiles.remove((x,y,z))
    else:
      flipped_tiles.add((x,y,z))
  return flipped_tiles

def part1(input):
  flipped_tiles = get_initial_flipped(input)
  return len(flipped_tiles)

def get_neighbours(point):
  neighbours = set()
  for dir,delta in deltas.items():
    # Do with zip?
    x = point[0] + delta[0]
    y = point[1] + delta[1]
    z = point[2] + delta[2]
    neighbours.add((x,y,z))
  return neighbours

def count_neighbours(flipped, point):
  neighbour_count = 0
  for neighbour in get_neighbours(point):
    if neighbour in flipped:
      neighbour_count += 1
  return neighbour_count

def part2(input):
  flipped = get_initial_flipped(input)
  for i in range(100):
    points_to_check = set()
    for point in flipped:
      points_to_check.add(point)
      for neighbour in get_neighbours(point):
        points_to_check.add(neighbour)
    next_gen = set()
    for point in points_to_check:
      neighbour_count = count_neighbours(flipped, point)
      if point in flipped and neighbour_count in [1,2]:
        next_gen.add(point)
      elif point not in flipped and neighbour_count == 2:
        next_gen.add(point)
    print(i, len(flipped))
    flipped = next_gen

  return len(flipped)

if __name__ == "__main__":
  with open(input_file) as f:
    data = f.read()
  print("Part 1: ", part1(data))
  print("Part 2: ", part2(data))

File: day-24/lobby_layout.go
package main

import (
	"fmt"
	"io/ioutil"
	"regexp"
	"strings"
)

const inputFile string = "./day-24/input.txt"

type point struct {
	x, y, z int
}

func (p point) add(d point) point {
	return point{p.x + d.x, p.y + d.y, p.z + d.z}
}

func (p point) getNeighbours() map[point]bool {
	neighbours := make(map[point]bool)
	for _, delta := range deltas {
		neighbours[p.add(delta)] = true
	}
	return neighbours
}

var deltas = map[string]point{
	"e":  {1, -1, 0},
	"ne": {1, 0, -1},
	"nw": {0, 1, -1},
	"w":  {-1, 1, 0},
	"se": {0, -1, 1},
	"sw": {-1, 0, 1},
}

var dirRegex = regexp.MustCompile("(ne|nw|se|sw|e|w)")

func processLine(line string) []string {
	return dirRegex.FindAllString(line, -1)
}

func getFlipped(input string) map[point]bool {
	flipped := make(map[point]bool)
	for _, line := range strings.Split(strings.TrimSpace(input), "\n") {
		point := point{0, 0, 0}
		directions := processLine(line)
		for _, dir := range directions {
			point = point.add(deltas[dir])
		}
		if _, found := flipped[point]; found {
			delete(flipped, point)
		} else {
			flipped[point] = true
		}
	}
	return flipped
}

func countNeighbours(p point, active map[point]bool) int {
	neighbourCount := 0
	for neighbour := range p.getNeighbours() {
		if _, found := active[neighbour]; found {
			neighbourCount++
		}
	}
	return neighbourCount
}

//Part1 - Solution to Part 1 of the puzzle
func Part1(input string) int {
	flipped := getFlipped(input)
	return len(flipped)
}

//Part2 - Solution to Part 2 of the puzzle
func Part2(input string) int {
	flipped := getFlipped(input)
	for i := 0; i < 100; i++ {
		pointsToCheck := make(map[point]bool)
		for point := range flipped {
			pointsToCheck[point] = true
			for neighbour := range point.getNeighbours() {
				pointsToCheck[neighbour] = true
			}
		}
		nextGen := make(map[point]bool)
		for point := range pointsToCheck {
			neighbourCount := countNeighbours(point, flipped)
			if _, found := flipped[point]; found {
				if neighbourCount == 1 {
					nextGen[point] = true
				}
			}
			if neighbourCount == 2 {
				nextGen[point] = true
			}
		}
		flipped = nextGen
	}
	return len(flipped)
}

func main() {
	b, err := ioutil.ReadFile(inputFile)
	if err != nil {
		fmt.Print(err)
	}
	data := string(b)

	fmt.Printf("Part 1: %d\n", Part1(data))
	fmt.Printf("Part 2: %d\n", Part2(data))
}
